Newsletters: add 154 (2021-06-23) (#592)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2021-06-23-newsletter.md

      bumping need to account for a malicious counterparty's ability to
      prevent setting the BIP125 signal.  If every transaction could be
      replaced, this wouldn't be a concern.

    - **Reducing transaction analysis opportunity** because transactions
      that opt-in to RBF look different onchain than transactions which
      don't.  Since most wallets consistently opt-in, or not, this
      provides evidence that surveillance companies can use in their
      attempts to identify who owns which bitcoins.  If every
      transaction was replaceable, there'd be no need to set the BIP125
      signal.

    This week, Antoine Riard [posted][riard rbf] a proposal to the Bitcoin-Dev
    mailing list for eventually changing Bitcoin Core's code to allow
    RBF for all transactions regardless of whether or not they set the
    BIP125 opt-in signal.  The idea was also discussed in the first
    transaction relay workshop [meeting][trw meeting].  Several meeting
    participants mentioned Bitcoin Core [PR #10823][bitcoin core #10823]
    as an alternative approach---it allows any transaction to be
    replaced, but only after the transaction had spent a certain amount

>     of time in a node mempool (originally proposed as six hours; latter

Reviewer: ```suggestion
    of time in a node mempool (originally proposed as six hours; later
```

    suggested to be 72 hours).

    Both Riard's email and meeting participants noted that any proposal
    for replacing transactions that don't contain a BIP125 opt-in signal
    requires feedback from any merchants who are currently depending on
    BIP125 behavior.  Optech encourages any such merchants to either

>     respond to mailing list thread or at least communicate any concerns

Reviewer: ```suggestion
    respond to the mailing list thread or at least communicate any concerns
```

  original transaction.  This opt-in behavior was proposed as a
  compromise between people who wanted to allow transaction replacement,
  such as for fee bumping or [additive payment batching][], and people
  who objected because allowing replacement simplifies building tools
  that defraud merchants who accept unconfirmed transactions as final.

    Over five years later, it appears very few merchants today are
    accepting unconfirmed transactions as final, and it's not clear how
    many of those that do are actually checking for the BIP125 opt-in
    signal and treating those transactions differently.  If no one is
    relying on BIP125 signals, then allowing every transaction to be
    replaceable could provide some advantages, such as:

    - **Simplifying analysis** for presigned transaction protocols (such
      as LN and [vaults][topic vaults]) where ideas for using RBF fee
      bumping need to account for a malicious counterparty's ability to
      prevent setting the BIP125 signal.  If every transaction could be
      replaced, this wouldn't be a concern.

    - **Reducing transaction analysis opportunity** because transactions

>       that opt-in to RBF look different onchain than transactions which

Reviewer 2: s/opt-in/opt in/ for the verb form, here and also on the next line 45

      don't.  Since most wallets consistently opt-in, or not, this
      provides evidence that surveillance companies can use in their
      attempts to identify who owns which bitcoins.  If every
      transaction was replaceable, there'd be no need to set the BIP125
      signal.

    This week, Antoine Riard [posted][riard rbf] a proposal to the Bitcoin-Dev
    mailing list for eventually changing Bitcoin Core's code to allow
    RBF for all transactions regardless of whether or not they set the
    BIP125 opt-in signal.  The idea was also discussed in the first
    transaction relay workshop [meeting][trw meeting].  Several meeting
    participants mentioned Bitcoin Core [PR #10823][bitcoin core #10823]
    as an alternative approach---it allows any transaction to be
    replaced, but only after the transaction had spent a certain amount
    of time in a node mempool (originally proposed as six hours; later
    suggested to be 72 hours).

>     Both Riard's email and meeting participants noted that any proposal

Reviewer 2: ```suggestion
    Both Riard's email and the meeting participants noted that any proposal
```

    for replacing transactions that don't contain a BIP125 opt-in signal

>     requires feedback from any merchants who are currently depending on

Reviewer 2: three micro-nits, feel free to ignore

s/requires/would require/

can omit "any", which is already used earlier in the sentence ("any proposal") and at the start of the next one

s/are currently depending/currently depend/

    BIP125 behavior.  Optech encourages any such merchants to either

>     respond to the mailing list thread or at least communicate any concerns

Reviewer 2: s/any such merchants/merchants/ and can omit "either" and "at least"

>     they have to Riard and other developers directly.

Author: I'm changing this but I do wish that we had a way of soliciting feedback from people who don't want to expose themselves to publicly controversy or ridicule.

  default commitment transaction format, adds support for using a pruned
  Bitcoin full node, allows receiving and sending payments using Atomic
  MultiPath ([AMP][topic multipath payments]), and increases LND's
  [PSBT][topic psbt] capabilities, among many other improvements and bug
  fixes.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1
repo], [Hardware Wallet Interface (HWI)][hwi repo],
[Rust Bitcoin][rust bitcoin repo], [BTCPay Server][btcpay server repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

- [Bitcoin Core #21365][] adds the ability for the wallet to create
  signatures for [taproot][topic taproot] spends---both keypath spends
  using only the P2TR public key and scriptpath spends using a
  [tapscript][topic tapscript].  The wallet can also sign for
  taproot-spending [PSBTs][topic psbt], but only if the wallet already
  has all the keypath or scriptpath information it needs.  The somewhat

>   related merged PR [#22156][bitcoin core #22156] only allows importing

Reviewer 2: somewhat-related

  that keypath and scriptpath information after taproot is active (block

>   709,632 on mainnet, but on test networks where taproot is already

Reviewer 3: See previous comment about using a site-wide alias for trb.

  enabled, importing may be used now).

- [Bitcoin Core #22144][] randomizes the order in which peers are serviced
  in the message handling thread. Bitcoin Core's message handling thread is

>   responsible for parsing and processing p2p messages from peers and for

Reviewer 2: s/p2p/P2P|peer-to-peer/ 

to avoid writing "message handling thread" twice in a row, perhaps join the first two sentences:

"...are serviced in Bitcoin Core's message handling thread, which is responsiblefor ..."

Reviewer 3: I've taken both suggestions. Thanks!

  sending messages to those peers. Previously, the message handling thread
  would service each peer round-robin _in the order in which the
  connections to those peers were first established_. This PR changes the
  logic so that, on each iteration of the message handling loop, the order in
  which peers are serviced is randomized. Peers are still serviced with the same
  frequency (each peer is serviced once per iteration), but any weaknesses or
  exploits that rely on a deterministic ordering of servicing peers are
  avoided.

- [Bitcoin Core #21261][] makes it easier to extend inbound connection
  diversity protection to more networks and then uses that framework to

>   add [I2P][] the list of diversity-protected networks.  Diversity

Reviewer 2: ```suggestion
  add [I2P][] to the list of diversity-protected networks.  Diversity
```
or shorter and clearer IMO,  "to extend eviction protection to I2P peers."

>   protection (often called eviction protection) allows a few peers with

Author: Yeah, I just made up "diversity protection" because I wasn't happy with "eviction protection".  Eviction protection tells readers *what* you're doing, but diversity protection tells them *why* you're doing it.

I agree I should rephrase to avoid mentioning the term as much, and thanks for the link reference.

>   protection (often called eviction protection) allows a few peers with

Reviewer 2: Yes, they are also using "protection" in opposite ways: protection from vs protection of.

  desirable characteristics to remain connected when Bitcoin Core is
  otherwise pruning low-quality connections.  Retaining a few
  connections to peers on anonymity networks is highly desirable both
  because it allows transaction creators to use those networks to hide
  their network identity and because the ability to receive blocks over
  those networks in addition to the regular Internet Protocol can
  prevent some types of [eclipse attacks][topic eclipse attacks].

- [Rust Bitcoin #601][] adds support for parsing [bech32m][topic bech32]
  addresses and requires that v1+ native segwit addresses be encoded with
  bech32m and not bech32.

- [BTCPay Server #2450][] makes generating [payjoin][topic
  payjoin]-compatible invoices the default when the user opts into using
  a hot wallet for receiving payments.  A button on the *create wallet*

>   screen allows the user to opt-out of payjoin receiving support.

Reviewer 2: (a) no hyphen for the verb form, (b) not sure if this is better but "payjoin receiving support" seems incorrect or awkward
```suggestion
  screen allows the user to opt out of this default setting.
```

Author: Blarg, sorry I suck at hyphen usage and you keep needing to correct me.  :-(

- [BTCPay Server #2559][] adds a separate screen for guiding the user

>   through their choices for how to sign transactions spending from their

Reviewer 2: ```suggestion
  through their choices for how to sign transaction spending from their
```

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [LND 0.13.0-beta][] is a new major release that improves feerate
  management by making [anchor outputs][topic anchor outputs] the
  default commitment transaction format, adds support for using a pruned
  Bitcoin full node, allows receiving and sending payments using Atomic
  MultiPath ([AMP][topic multipath payments]), and increases LND's
  [PSBT][topic psbt] capabilities, among many other improvements and bug
  fixes.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1
repo], [Hardware Wallet Interface (HWI)][hwi repo],
[Rust Bitcoin][rust bitcoin repo], [BTCPay Server][btcpay server repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

- [Bitcoin Core #21365][] adds the ability for the wallet to create
  signatures for [taproot][topic taproot] spends---both keypath spends

>   using only the P2TR public key and scriptpath spends using a
>   [tapscript][topic tapscript].  The wallet can also sign for

Reviewer: I would have taken "tapscript" to refer to the Script version (the set of opcodes) allowed in a Tapleaf. Maybe one of these:
```suggestion
  using only the P2TR public key and scriptpath spends using
  [tapscript][topic tapscript].  The wallet can also sign for
```

```suggestion
  using only the P2TR public key and scriptpath spends using the
  [Taptree][topic Taproot].  The wallet can also sign for
```

Author: I'm not 100% sure, but I think BIP342 is using "tapscript" the same way I am (emphasis in original):

> The script as defined in BIP341 (i.e., the penultimate witness stack element after removing the optional annex) is called the **tapscript** and is decoded into opcodes [...]

I don't have a strong feeling here.  If you do, please say so and I'll make one of your suggested changes.

Reviewer: Nope, all good. I didn't realize it was also used like this!
